In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Assigning the project folder patht to path variable
path = r'C:\Users\Ruth Kassaye\Documents\CareerFoundery\DataImmersion\Achievement_4'

In [3]:
# Printing path variable to check if the folder path is correctly printed out
path

'C:\\Users\\Ruth Kassaye\\Documents\\CareerFoundery\\DataImmersion\\Achievement_4'

In [4]:
# Importing orders.csv file
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [5]:
# Printing df_ords dataframe
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [6]:
# Importing products.csv file
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

## Dropping Column

In [7]:
#Dropping eval_set column from orders.csv
df_ords.drop (columns= ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

### Assigning/creating a new dataframe once I have dopped eval_set column/variable

In [8]:
df_ords_new = df_ords.drop (columns= ['eval_set'])

In [9]:
# Printing the new dataframe(df_ords_new)
df_ords_new

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

## Frequency of a given column
- For example: days_since_prior_order column

In [10]:
# To know the frequency of 'days_since_prior_order' variable. dropna= False means telling panda to also count null values
df_ords ['days_since_prior_order'].value_counts (dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

- For example: order_number column

In [11]:
# To know the frequency of 'order_number' variable.
df_ords ['order_number'].value_counts(dropna = False)

1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: order_number, Length: 100, dtype: int64

## Renaming columns/variables
-Renaming **'order_dow'** to **'orders_day_of_week'** of the df_order dataframe without overwriting the data frame; using **inplace=True function**

In [12]:
# Renaming 'order_dow'to 'orders_day_of_week' of the df_order dataframe
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [13]:
# Printing the df_ords dataframe to check panda has indeed changed my the variable/column title
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [14]:
# Printing descriptive statistics of df_order dataframe
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## Changing data type
 - changing data type of a single column

In [15]:
# Changing order_id column datatype from numeric(datatype in the original dataframe) to string (str)
df_ords['order_id'] = df_ords['order_id'].astype('str')

- changing data type of multiple columns (**both order_id and user_id**)

In [16]:
# If I want to chanage the data type of more than two columns. First I need to list the columns I want to operate the change data type operation, in this case order_id and user_id
coulmn_list = ['order_id', 'user_id']

In [17]:
# changing the data type of columns using astype function
df_ords[coulmn_list]= df_ords[coulmn_list].astype ('str')

In [18]:
# printing out data type of df_order dataframe to check the change of data type of order_id and user_id. It is object and means string
df_ords.dtypes

order_id                   object
user_id                    object
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [19]:
# Also I can check the data tupe of columns using info function
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


- changing column data type using apply function

In [20]:
#using apply function
df_ords['order_id'].dtypecoulmn_list= df_ords[coulmn_list].apply (str)

In [21]:
# Printing the data type of df_order to check if the apply function has changed the data type I did in the above step
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


In [22]:
# Printing order_id datatype to check if panda has changed the datatype to str. 0 in panda refers to string
df_ords['order_id'].dtype

dtype('O')

In [23]:
df_ords[coulmn_list].dtypes

order_id    object
user_id     object
dtype: object

## Descriptive Statistics Analysis 

In [24]:
df_ords.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count  3.421083e+06        3.421083e+06       3.421083e+06   
mean   1.715486e+01        2.776219e+00       1.345202e+01   
std    1.773316e+01        2.046829e+00       4.226088e+00   
min    1.000000e+00        0.000000e+00       0.000000e+00   
25%    5.000000e+00        1.000000e+00       1.000000e+01   
50%    1.100000e+01        3.000000e+00       1.300000e+01   
75%    2.300000e+01        5.000000e+00       1.600000e+01   
max    1.000000e+02        6.000000e+00       2.300000e+01   

       days_since_prior_order  
count            3.214874e+06  
mean             1.111484e+01  
std              9.206737e+00  
min              0.000000e+00  
25%              4.000000e+00  
50%              7.000000e+00  
75%              1.500000e+01  
max              3.000000e+01

### Importing departments.csv

In [25]:
# Importing df_dep
df_dep = pd.read_csv(os.path.join(path,'02 Data', 'Original Data','departments.csv'))

In [26]:
# Printing df_dep
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

## Transposing data
- Transposing means turning my dataframe’s rows into columns and vise versa.
- **Why trasposing?**: sometimes dataframes could be wide. See for instance df_dep dataframe imported above.There’s only 1 row and 22 columns, making the dataframe incredibly wide and incredibly short. This is what “wide format” looks like.

## Step 1:  Trasposing the imported dataframe above (**df_dep**) and assigning a new dataframe to the trasposed dataframe(**df_dep_t**)

In [27]:
# Trasposing df_dep dataframe and assigning for the trasposed dataframe a new dataframe, df_dept_t
df_dep_t = df_dep.T

In [28]:
# Printing the entire rows of the df_dept_t dataframe
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## Step 2: Creating an index
- This function is the first step to remove the 0 header. See above the trasposed dataframe. We don't want to be presented like this.

In [29]:
# adding and index in the header. Note that I have two rows of headers. So I need to have only one header
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

## Step 3: Creating a new header 
- This is the header I want to have in my **df_dep_t** dataframe
- **iloc[0]** means I am telling panda to select index 0 or the first row as my new header. In the above step I had two headers. So I am choosing here which of the headers to be my real header

In [30]:
# Creating a new header (new_header).   
new_header = df_dep_t.iloc[0]

In [31]:
new_header

0    department
Name: department_id, dtype: object

## Step 4: Removing the first row from the df_dep_t dataframe and copy the dataframe into a new daraframe
- Now that I have selected index 0 to be my new_header. I will copy every rows below my new heade starting from **index 0** or **first row** into a new dataframe **(df_de_t_new)**

In [32]:
# Tellign Python to copy everything starting from first row/index 0 
df_dep_t_new = df_dep_t[1:]

In [33]:
# Printing out the df_dep_t_new dataframe. Note that this looks exactly like the old dataframe only without the first row. I should now set my new_header variable as the new header for my df_dep_t_new dataframe
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

## Step 5: See the table above, it is exactly like the original dataframe, having 0 as a header. So, at this step, I should replace 0 header with my new_header I have created in Step 3 above

- Hence the below function will assign the **new_header** to **df_dep_t_new** dataframe.

In [34]:
# Telling panda to take the values stored in the new_header variable and put them in the df_dep_t_new dataframe as the names of its columns.
df_dep_t_new.columns = new_header

## Step 6: Printing the final Trasposed Dataframe!

In [35]:
# Printing out df_df_dep_new to check if my new_header is integrated in the df_dep_t_new dataframe
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## Creating Data  Dictionaries  for the 'department_id' column

In [36]:
# creating a data dictionary for the values in the "department_id" column of the df_dep_t_new dataframe
data_dict = df_dep_t_new.to_dict('index')

In [37]:
# calling my new data_dict variable
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [38]:
# printing the first 5 rows of df_prods dataframe. 
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [39]:
# Row 0 has a "department_id" of 19. I can use my data dictionary to determine what this stands for with the following command:
print(data_dict.get('19'))

{'department': 'snacks'}


## Ask suzan what if I want to select more ids and see their meaning using my data dictionary
- Below is what I have tried, but I was not successful 

print(data_dict.get[Dept_id_list])

In [40]:
Dept_id_list = [19,4,13]

## Subsetting

In [41]:
# Printing department_id 19 from the df_prods dataframes (this process is called subsetting)
df_prods[df_prods['department_id'] ==19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [42]:
# saving the results above, values of department_id ==19, into a new dataframe
df_snack= df_prods[df_prods['department_id']==19]

In [43]:
# Calling df_snack
df_snack

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [44]:
# # saving the results above, values of department_id ==17,18,19, into a new dataframe df_snack_2
df_snack_2= df_prods.loc[df_prods['department_id'].isin([17,18,19])]

In [45]:
df_snack_2

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
13             14                 Fresh Scent Dishwasher Cleaner        74   
14             15                       Overnight Diapers Size 6        56   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
...           ...                                            ...       ...   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49672       49668                 Apple Cinnamon Scented Candles       101   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
13                17     6.5  
14                18    11.2  
15                19     5.2  
24                19     1.9  
...              ...     ...  
49669             19     6.9  
49670             19     5.2  
49672             17     5.6  
49675             19     3.0  
49683             19     6.0  

[10430 rows x 5 columns]

# Task 4.4

### Q2. Find another identifier variable in the df_order dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.
- **Answer**: I have done for the user_id column as well. Please see the Changing data type section

## Q3. Look for a variable in your df_order dataframe with an unintuitive name and change its name without overwriting the data frame.
- **Answer**: I have done above; please see the Renaming columns/variables section


## Q4. Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [46]:
df_ords ['order_hour_of_day'].value_counts(dropna = False)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

**Answer**: as the frequency table above shows, the busies hour is 10 with a total of 288,418 orders.

   -**Reflection**: I was not certain if the hour is pm or am. The dataset doesn't indicate this.

## Q5. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.
- Please see the Data Dictionary section. I have already created Data Dictionary for the department_id

In [47]:
print(data_dict.get('4'))

{'department': 'produce'}


**Answer**: the department is 'produce'

## Q6. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

- I know that the department_id is located in the df_prods dataframe. Hence the dataframe would be my whole dataframe where I am going to map the department_id that has 14 values. 
- Also, from my department_id data dictionary I know that breakfast item sales has an id of 14.
- See the below code that I wrote to get more retrieve information about department_id ==14(breakfast sales) from the df_prods dataframe

In [48]:
df_prods[df_prods['department_id'] == 14]

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

## Q7. They’d also like to see details about customers who might be throwing dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

- From my department_id dictionary, I know that alcohol's id is 5, deli's id is 20, beverages's id is 7, and meat/seafood's id is12
- **isin** function is  more handy here to select more values that has multiple department_ids at the same time 

In [49]:
# Subsetting slaes for dinner parties as df_dinner_parties dataframe
df_dinner_parties = df_prods.loc[df_prods['department_id'].isin([5,20,7,12])]

In [50]:
# Printing the df_dinner_parties dataframe to check the subsetting I did in the above step
df_dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

## Q8. It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

- df_dinner_parties has 7650 rows, see the table above 

## Q9. Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [51]:
df_user_id_1 = df_ords[df_ords['user_id'].isin(['1'])]

# I can also use: df_user_id_1 = df_order.loc[df_order['user_id']=='1']

**N.B.** 
- I put 1 in quotation mark as user_id is now string. I have changed both the user_id and order_id columns data type from numeric to string(str) data type. Please see above the Changing data type section. 

In [52]:
df_user_id_1

order_id user_id eval_set  order_number  orders_day_of_week  \
0   2539329       1    prior             1                   2   
1   2398795       1    prior             2                   3   
2    473747       1    prior             3                   3   
3   2254736       1    prior             4                   4   
4    431534       1    prior             5                   4   
5   3367565       1    prior             6                   2   
6    550135       1    prior             7                   1   
7   3108588       1    prior             8                   1   
8   2295261       1    prior             9                   1   
9   2550362       1    prior            10                   4   
10  1187899       1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

## Q10. You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [53]:
# Printing descriptive statistics about of df_user_id_1 dataframe
df_user_id_1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count     11.000000           11.000000          11.000000   
mean       6.000000            2.636364          10.090909   
std        3.316625            1.286291           3.477198   
min        1.000000            1.000000           7.000000   
25%        3.500000            1.500000           7.500000   
50%        6.000000            3.000000           8.000000   
75%        8.500000            4.000000          13.000000   
max       11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

## Q11. Check the organization and structure of your notebook. Be sure to include section headings and code comments.

- **Answer**: Done

## Q12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [54]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

## Q13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [55]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'departments_wrangled.csv'))